src/tactic/compute_degree.lean

```diff
@@ -172,8 +298,182 @@ by compute_degree_le
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, m], lc],
+  c_c ← assert nn co_eq_co,
+  interactive.swap,
+  if t_is_eq then refine ``(eq.trans %%c_c _) else refine ``(ne_of_eq_of_ne %%c_c _),
+  try $ tactic.clear c_c,
+  interactive.swap,
+  resolve_coeff
+
+section parsing
+setup_tactic_parser
+
+/--
+`reduce_coeff` takes a polynomial `f` as an input and produces a hypothesis of the form
+`c_c : f.coeff n = <simplified expression>`, where the simplified expression is obtained via the
+`get_lead_coeff/resolve_coeff` pair.
+-/
+meta def reduce_coeff (fp : parse texpr) : tactic unit :=
+do
+  f ← to_expr ``(%%fp) tt ff,
+  exp_deg ← guess_degree' f,
+  `(@polynomial %%R %%_) ← infer_type f,
+  c ← mk_instance_cache R,
+  (c, lc) ← get_lead_coeff c f,
+  nn ← get_unused_name "c_c",
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, `(exp_deg)], lc],
+  assert nn co_eq_co,
+  resolve_coeff
+
+end parsing
+
+/--  `compute_degree` tries to close goals of the form `f.(nat_)degree = d`.  It converts the
+goal to showing that
+* the degree is at most `d`, calling `compute_degree_le` to solve this case;
+* the coefficient of degree `n` is non-zero, calling `simp_coeff` to simplify this goal.
+Unless the polynomial is particularly complicated, `compute_degree` with either succeed of leave
+a simpler goal to prove.
+ -/
+meta def compute_degree : tactic unit := focus $ do
+t ← target >>= (λ f, whnf f reducible),
+match t with
+  -- the `degree` match implicitly assumes that the `nat_degree` is strictly positive
+| `(    degree %%_ = %%_) := refine ``((degree_eq_iff_nat_degree_eq_of_pos _).mpr _) >> rotate
+| `(nat_degree %%_ = %%_) := single_term_suggestions
+| _ := fail "Goal is not of the form\n`f.nat_degree = d` or `f.degree = d`"
+end,
+`(nat_degree %%pol = %%degv) ← target |
+  fail "Goal is not of the form\n`f.nat_degree = d` or `f.degree = d`",
+deg ← guess_degree' pol,
+degvn ← eval_expr' ℕ degv,
+guard (deg = degvn) <|>
+( do ppe ← pp deg, ppg ← pp degvn,
+  fail sformat!("'{ppe}' is the expected degree\n'{ppg}' is the given degree\n") ),
+refine ``(le_antisymm _ (le_nat_degree_of_ne_zero _)),
+focus' [compute_degree_le, simp_coeff]
+
+/--  `prove_monic` tries to close goals of the form `monic f`.  It converts the
+goal to showing that
+* the degree is at most `d`, calling `compute_degree_le` to solve this goal;
+* the coefficient of degree `d` equals `1`, calling `simp_coeff` to simplify this goal.
+Unless the polynomial is particularly complicated, `prove_monic` will either succeed or leave
+a simpler goal to prove.
+ -/
+meta def prove_monic : tactic unit := focus $ do
+`(monic %%pol) ← target >>= (λ f, whnf f reducible) | fail"Goal is not of the form `monic f`",
+deg ← guess_degree' pol,
+deg ← to_expr deg.to_pexpr tt ff,
+refine ``(monic_of_nat_degree_le_of_coeff_eq_one %%deg _ _),
+focus' [compute_degree_le, simp_coeff],
+try reflexivity
 
```

Comment:
Could you add tactic docs for the new user-facing tactics?

Reply:
I added tactic docs for `compute_degree, prove_monic, simp_coeff`.  If you think that `simp_coeff` should not be in the interactive namespace, let me know.